### PR TITLE
Build profiles and reproducible builds

### PR DESCRIPTION
**Pull request by cmaglie**
This RFC aims to propose new augmentations to the current library management and a build system that can be entirely transparent to the user and retains backward compatibility yet allowing more flexibility for the future.

### CODE REVIEW

File: RFCs/0003-build-profiles.md [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+      - <LIB_NAME> (<LIB_VERSION>)
+
+...more profiles here...
+
+default_profile: <DEFAULT_PROFILE_NAME>
+```
+
+We have a `profiles:` section containing all the profiles. Each field is self-explanatory, in particular:
+
+- `<PROFILE*NAME>` is the profile identifier, it’s a user-defined field, and the allowed characters are alphanumerics, underscore `_`, dot `.`, and dash `-`
```
Comment (per1234) [inline RFCs/0003-build-profiles.md:76]:
```suggestion
- `<PROFILE_NAME>` is the profile identifier, it’s a user-defined field, and the allowed characters are alphanumerics, underscore `_`, dot `.`, and dash `-`
```

File: RFCs/0003-build-profiles.md [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+arduino-cli compile [sketch]
+```
+
+will, instead, trigger a profile-based build using the default profile indicated in the `sketch.yaml`.
+
+## Vendoring libraries with custom modifications
+
+We will add the possibility to add custom libraries directly in the sketch using the `libraries` subdirectory in the sketch root folder.
+
+A typical usage scenario is when the sketch needs a library that is not part of the library manager, or if the sketch needs a library with some customizations that are not available in the upstream release.
```
Comment (per1234) [inline RFCs/0003-build-profiles.md:159]:
```suggestion
A typical usage scenario is when the sketch needs a library that is not offered for installation from Library Manager, or if the sketch needs a library with some customizations that are not available in the upstream release.
```

File: RFCs/0003-build-profiles.md [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+To accomplish this the idea is to put the custom libraries' source code directly in the sketch `libraries` subdirectory (exactly as we do for the globally installed libraries in the `libraries` subdirectory of the sketchbook root folder). For example:
```
Comment (per1234) [inline RFCs/0003-build-profiles.md:161]:
```suggestion
To accomplish this the idea is to put the custom libraries' source code directly in the `libraries` subdirectory of the sketch (exactly as we do for the globally installed libraries in the `libraries` subdirectory of the sketchbook root folder). For example:
```

File: RFCs/0003-build-profiles.md [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+These incompatibilities might arise due to newer versions of a Platform, Tool, or Library.
+
+## Problem
+
+An Arduino project is known as Sketch. It comprises its main .ino file plus optional local source code (in the same folder) and an optional extra src folder containing additional sources if required.
+
+The sketch compiles against a set of globally installed libraries and platforms. When said library and/or part of the hardware platform is updated, sometimes introducing some breaking changes, that sketch might not compile anymore or behave differently.
+
+Currently, Arduino guarantees no portability and reproducibility of a build at a later time unless exact instructions are provided by the developer to the final user, such as the required platform, including version and possibly a 3rd party platform URL to download and install the platform, or all the required libraries including the exact version of each one.
```
Comment (per1234) [inline RFCs/0003-build-profiles.md:15]:
```suggestion
Currently, the only way to guarantee portability and reproducibility of a sketch build at a later time is for the developer to provide instructions to the final user for the installation of the sketch dependencies, including the exact versions of the boards platform and libraries.
```
I think this wording makes the meaning slightly more clear.

**Review by per1234 — CHANGES_REQUESTED** [review]

**Comment by ladyada** [comment]
Hi please add the ability to dynamically add -D define lines on the sketch compilation thru the yaml!

**Comment by cmaglie** [comment]
> Hi please add the ability to dynamically add -D define lines on the sketch compilation thru the yaml!

Yep, we are going to add this feature later: we are keeping this version of the RFC as simple as possible so we can get it implemented ASAP and extend it later as we need.

(adding defines to profiles is tracked here: https://github.com/arduino/tooling-rfcs/issues/9)

**Review by per1234 — APPROVED** [review]

**Pull request merged**